validation/open_boundaries/nonhydrostatic_channel.jl

```diff
@@ -0,0 +1,46 @@
+using Oceananigans
+
+using Oceananigans.BoundaryConditions: OrlanskiBoundary
+using Oceananigans.TimeSteppers: QuasiAdamsBashforth2TimeStepper
+
+grid = RectilinearGrid(topology = (Bounded, Periodic, Bounded), size = (64, 32, 16), extent = (40, 20, 5), halo = (5, 5, 5))
+
+timestepper = QuasiAdamsBashforth2TimeStepper(grid, (:T, ))
+
+u_boundaries = FieldBoundaryConditions(west = OpenBoundaryCondition(.1), 
+                                       east = OpenBoundaryCondition(OrlanskiBoundary(timestepper.Gⁿ.u)))
+
+v_boundaries = FieldBoundaryConditions(west = OpenBoundaryCondition(OrlanskiBoundary(timestepper.Gⁿ.v)), 
+                                       east = OpenBoundaryCondition(OrlanskiBoundary(timestepper.Gⁿ.v)))
+
+w_boundaries = FieldBoundaryConditions(west = OpenBoundaryCondition(OrlanskiBoundary(timestepper.Gⁿ.w)), 
+                                       east = OpenBoundaryCondition(OrlanskiBoundary(timestepper.Gⁿ.w)))
+
+T_boundaries = FieldBoundaryConditions(west = OpenBoundaryCondition((y, z, t) -> ifelse(z < -4, 3, 0)), 
```

Comment:
for tracers you need to specify `ValueBoundaryCondition`. Open will change the value at 1 but you want to change the value outside the domain for temperature

Reply:
Hm no I think these are different. `ValueBoundaryCondition` sets the tracer value _on_ the boundary. Whereas for an open boundary condition, we are specifying the tracer values in a cell that lies outside the boundary.

Reply:
`OpenBoundaryCondition` sets the value at `1` which for the tracer is inside the domain, not a boundary. `Value` should be used to set the boundary for tracers (another option is to redefine `Open` for centered fields).

Reply:
`Value` doesn't set the tracer value --- it computes the halo point by extrapolating from the interior point through the boundary point.

Reply:
`OpenBoundaryCondition` is supposed to fill halo regions. If it doesn't that's a bug.

Reply:
That is what I was saying, to set a boundary for a tracer, `Value` is what we need. `Open` is for velocities. If we use `Open` for both `T` and `u` we end up with inlet conditions at different points.

Reply:
Then we have to fix `Open` to be "location-aware".

`Value` won't work. `Value` determines the halo value by extrapolating across the boundary. This is _not_ what we want for an open boundary condition. Think of an open boundary as connecting to another domain that is adjacent to the prognostic domain. To use an open boundary we want to specify the halo state directly. The reconstructed value of a tracer should be the average of the specified halo state and the interior state. This is _different_ from `Value`. `Value` will _not_ work for open boundaries.

Reply:
`OpenBoundaryCondition` is more like a Periodic boundary condition. In principle we should be able to specify all values of the halo state and use a high-order advection scheme all the way up to the boundary, for example.